File: cartunez-api/app/models/vehicle.py
"""Vehicle-related models for make, model, year, and variant."""

import uuid
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base


class VehicleMake(Base):
    """Vehicle manufacturer (e.g., Toyota, Honda)."""

    __tablename__ = "vehicle_makes"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    name: Mapped[str] = mapped_column(String(100), unique=True, nullable=False, index=True)
    slug: Mapped[str] = mapped_column(String(100), unique=True, nullable=False, index=True)
    logo_url: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    models: Mapped[list["VehicleModel"]] = relationship(
        "VehicleModel", back_populates="make", cascade="all, delete-orphan"
    )

    def __repr__(self) -> str:
        return f"<VehicleMake {self.name}>"


class VehicleModel(Base):
    """Vehicle model under a make (e.g., Camry, Civic)."""

    __tablename__ = "vehicle_models"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    make_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("vehicle_makes.id", ondelete="CASCADE"), nullable=False
    )
    name: Mapped[str] = mapped_column(String(100), nullable=False, index=True)
    slug: Mapped[str] = mapped_column(String(100), nullable=False, index=True)
    body_type: Mapped[str | None] = mapped_column(String(50), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    make: Mapped["VehicleMake"] = relationship("VehicleMake", back_populates="models")
    years: Mapped[list["VehicleYear"]] = relationship(
        "VehicleYear", back_populates="model", cascade="all, delete-orphan"
    )

    def __repr__(self) -> str:
        return f"<VehicleModel {self.name}>"


class VehicleYear(Base):
    """Vehicle model year."""

    __tablename__ = "vehicle_years"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    model_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("vehicle_models.id", ondelete="CASCADE"), nullable=False
    )
    year: Mapped[int] = mapped_column(Integer, nullable=False, index=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    model: Mapped["VehicleModel"] = relationship("VehicleModel", back_populates="years")
    variants: Mapped[list["VehicleVariant"]] = relationship(
        "VehicleVariant", back_populates="vehicle_year", cascade="all, delete-orphan"
    )

    def __repr__(self) -> str:
        return f"<VehicleYear {self.year}>"


class VehicleVariant(Base):
    """Specific vehicle variant with engine/trim details."""

    __tablename__ = "vehicle_variants"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    vehicle_year_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("vehicle_years.id", ondelete="CASCADE"), nullable=False
    )
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    engine: Mapped[str | None] = mapped_column(String(100), nullable=True)
    transmission: Mapped[str | None] = mapped_column(String(50), nullable=True)
    fuel_type: Mapped[str | None] = mapped_column(String(50), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    vehicle_year: Mapped["VehicleYear"] = relationship("VehicleYear", back_populates="variants")

    def __repr__(self) -> str:
        return f"<VehicleVariant {self.name}>"

File: backend/cartunez-medusa/src/models/vehicle.ts
import { Entity as Entity_, Column, PrimaryGeneratedColumn, ManyToOne, OneToMany, Index, JoinColumn, CreateDateColumn, UpdateDateColumn } from "typeorm";

@Entity_("vehicle_make")
export class VehicleMake {
  @PrimaryGeneratedColumn("uuid")
  id: string;

  @Column({ type: "varchar", length: 100, unique: true })
  name: string;

  @Column({ type: "varchar", length: 50, nullable: true })
  country: string;

  @Column({ type: "boolean", default: true })
  is_active: boolean;

  @CreateDateColumn({ type: "timestamptz" })
  created_at: Date;

  @UpdateDateColumn({ type: "timestamptz" })
  updated_at: Date;

  @OneToMany(() => VehicleModel, (model) => model.make)
  models: VehicleModel[];
}

@Entity_("vehicle_model")
export class VehicleModel {
  @PrimaryGeneratedColumn("uuid")
  id: string;

  @Column({ type: "varchar", length: 150 })
  name: string;

  @Column({ type: "uuid" })
  @Index()
  make_id: string;

  @ManyToOne(() => VehicleMake, (make) => make.models, { onDelete: "CASCADE" })
  @JoinColumn({ name: "make_id" })
  make: VehicleMake;

  @Column({ type: "varchar", length: 50, nullable: true })
  body_type: string;

  @Column({ type: "boolean", default: true })
  is_active: boolean;

  @CreateDateColumn({ type: "timestamptz" })
  created_at: Date;

  @UpdateDateColumn({ type: "timestamptz" })
  updated_at: Date;

  @OneToMany(() => VehicleYear, (year) => year.model)
  years: VehicleYear[];
}

@Entity_("vehicle_year")
export class VehicleYear {
  @PrimaryGeneratedColumn("uuid")
  id: string;

  @Column({ type: "int" })
  year: number;

  @Column({ type: "uuid" })
  @Index()
  model_id: string;

  @ManyToOne(() => VehicleModel, (model) => model.years, { onDelete: "CASCADE" })
  @JoinColumn({ name: "model_id" })
  model: VehicleModel;

  @Column({ type: "boolean", default: true })
  is_active: boolean;

  @CreateDateColumn({ type: "timestamptz" })
  created_at: Date;

  @UpdateDateColumn({ type: "timestamptz" })
  updated_at: Date;

  @OneToMany(() => VehicleVariant, (variant) => variant.year)
  variants: VehicleVariant[];
}

@Entity_("vehicle_variant")
export class VehicleVariant {
  @PrimaryGeneratedColumn("uuid")
  id: string;

  @Column({ type: "varchar", length: 200 })
  name: string;

  @Column({ type: "uuid" })
  @Index()
  year_id: string;

  @ManyToOne(() => VehicleYear, (year) => year.variants, { onDelete: "CASCADE" })
  @JoinColumn({ name: "year_id" })
  year: VehicleYear;

  @Column({ type: "varchar", length: 50, nullable: true })
  engine_type: string;

  @Column({ type: "varchar", length: 50, nullable: true })
  fuel_type: string;

  @Column({ type: "varchar", length: 50, nullable: true })
  transmission: string;

  @Column({ type: "boolean", default: true })
  is_active: boolean;

  @CreateDateColumn({ type: "timestamptz" })
  created_at: Date;

  @UpdateDateColumn({ type: "timestamptz" })
  updated_at: Date;
}

@Entity_("product_vehicle_compatibility")
export class ProductVehicleCompatibility {
  @PrimaryGeneratedColumn("uuid")
  id: string;

  @Column({ type: "uuid" })
  @Index()
  product_id: string;

  @Column({ type: "uuid" })
  @Index()
  vehicle_variant_id: string;

  @ManyToOne(() => VehicleVariant, { onDelete: "CASCADE" })
  @JoinColumn({ name: "vehicle_variant_id" })
  vehicle_variant: VehicleVariant;

  @Column({ type: "varchar", length: 20, nullable: true })
  fitment_type: string; // "exact", "universal", "cross-reference"

  @Column({ type: "text", nullable: true })
  notes: string;

  @CreateDateColumn({ type: "timestamptz" })
  created_at: Date;
}
